[WIP] First pieces of Beam Sync phase 1

carver opened the pull request:
### What was wrong?

Some setup for #1225

### How was it fixed?

It's not. Just getting a rough start on introducing the new subprotocol to broadcast (and receive) the node hashes in the witness metadata.

~Some of it references APIs that probably won't be available until #1189 is merged.~
It also depends on things that aren't merged into py-evm yet. (Actually, they might not even be in a PR yet)

This is completely untested, just sharing early.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://images.fineartamerica.com/images-medium-large-5/bull-dog-and-the-fire-hydrant-standoff-linda-matlow.jpg)

carver committed:
Add Firehose subprotocol, hacked as ETH-like peer

The only job currently is to broadcast witness matadata: the set of node
hashes needed to prove the data read by a block.

carver committed:
Make the default peer firehose-enabled

carver committed:
move witness broadcaster to witness module

carver committed:
Prevent rebroadcasting witness to sending peers

If a peer sent us a witness, they clearly don't need one from us.

carver committed:
Remove some firehose API cruft

carver committed:
Swap out backfiller for witness collector

The Beam downloader needs a service that tracks the 'queen' peer. The
backfiller was doing a pretty terrible job at backfilling, and witness
collection should be good at tracking the queen. So this replaces the
backfiller with the witness collector, for queen tracking. (And also
for... collecting witnesses)

carver committed:
Broadcast event when witness is locally built

carver committed:
Witness collect & broadcast: renames & fixups

carver committed:
add event for witness broadcast

carver committed:
louder firehose logs about status broadcast

carver commented on trinity/sync/beam/witness.py:
```diff
@@ -0,0 +1,353 @@
+
+    async def _collect_peer_witness_metadata(self) -> None:
+        while self.is_operational:
+            peer, cmd, msg = await self.wait(self.msg_queue.get())
```
@pipermerriam What would be the "new API" way of doing this? I want to collect all messages of a certain type, as they come in, across all peers. (Notice that this class is only listening to the one message: `NewBlockWitnessHashes`)

I'm thinking it's something like: attach a custom behavior at runtime to every peer. Roughly:

```py
    def register_peer(self, peer: BasePeer) -> None:
        super().register_peer(peer)
        peer.connection.add_behavior(NewBlockWitnessHashes, self._handle_new_witness)

    def _handle_new_witness(self, command: NewBlockWitnessHashes):
        # do thing with the command
```

pipermerriam replied:
Hrm, we don't yet have a `PeerSubscriber` replacement which is the only current way to listen across all peers.  So yes, best way is to use the peer subscriber API to get access to each peer as it connects and then attach a `CommandHandler` for the `NewBlockWitnessHashes` to each peer as they come online.  I would not recommend using the `PeerSubscriber` APIs that are for subscribing to commands because those are probably about ready to be removed so make sure you use the `p2p.logic.CommandHandler` APIs for that part and only use the `PeerSubscriber` APIs to get access to the connected/disconnected events.

pipermerriam replied:
Here's an equivalent from the `ETH` protocol that listens for `NewBlock`

https://github.com/ethereum/trinity/blob/39a2d1eddeff0b9313677b53037b48ebcd54c527/trinity/protocol/eth/api.py#L26-L41

carver replied:
> make sure you use the `p2p.logic.CommandHandler` APIs for that part and only use the `PeerSubscriber` APIs to get access to the connected/disconnected events.

How do I find out when a new peer becomes available, so I can attach the new behavior using `add_behavior()`?

pipermerriam replied:
You would still use this API: https://github.com/ethereum/trinity/blob/86e43338e43d0130d471ef76b7db61cc85bc3c15/p2p/peer.py#L348

carver replied:
Hm, the more I read through the CommandHandler stuff the more I think it makes it difficult to add runtime handlers. If I do something like `peer.eth_api.add_child_behavior()` when I get a `register_peer` call, then I think this code will have already run:

https://github.com/ethereum/trinity/blob/992226b5036448b3e30f9638a6da73d1471e5a98/p2p/logic.py#L77-L89

So I think the *new* child behavior will be ignored, because the `Application` has already been applied to the connection. Do I have that right?

Maybe we need a new API on the peer (or connection or ?) to enable adding handlers at runtime.

pipermerriam replied:
I'm all for improvements to these APIs.  They were developed in a semi-vacuum and definitely have room to be improved based on real use cases.

For the specific issue you are referencing, I think you are correct that using `add_child_behaviour()` would not work as expected.  I'm inclined towards making `Application` reject any calls to `add_child_behavior()` once it has already been attached to a connection.

carver committed:
Fix up broken imports

carver committed:
core fixups

carver committed:
flake8 cleanup 1

carver committed:
mypy fixups

carver committed:
Flake8 fixups

carver committed:
remove unused SnappyCommand

carver committed:
update-for-new-queen-split

carver committed:
mypy and flake8 fixups

carver committed:
lint fixups and use StatelessImportDone instead of new event

carver committed:
Clean up some warnings and more clean service exit

Had some BeamBlockImporter._preview_address_load() warnings like:
Task was destroyed but it is pending!

carver committed:
Fix some delayed and stuck cancellations

carver committed:
fix type issues that only show up in CI :(

carver committed:
mypy fixups

carver commented on trinity/sync/beam/importer.py:
```diff
@@ -360,29 +362,67 @@ def _broadcast_import_complete(
+                f"Value is not a hash, because it's only {len(next_hash)} long: {next_hash}"
+            )
+        else:
+            yield cast(Hash32, next_hash)
```
I kind of like having it clear that this is only a type assertion and not an object instantiation, but only a mild preference.

cburgdorf replied:
The thing about the `cast` is that it will allow *anything* to be passed. Think of it as just turning to  `mypy` saying *"Get away mypy, trust me, I know what I'm doing"*. So you can write `cast(Hash32, "foo")` and it will happily accept it even though it is obviously wrong and will lead to an error at runtime.

On the other hand you can not write `Hash32("foo")`. Mypy will reject this with.

>"Hash32" has incompatible type "str"; expected "bytes"

carver committed:
Fixup protocol class attribute names

carver committed:
more-stuff

carver committed:
moar - I promise to clean this up before review

carver committed:
Add Firehose proto to ETHPeer

carver committed:
remove special firehose peer

carver committed:
show early message when block witness arrives

carver committed:
Fix to display extra stats on non-firehose peers

carver committed:
revert to master version of peer.cancel()

carver committed:
Quiet some *really* noisy errors on exit

carver committed:
Don't need two warnings about firehose peer join & leave

carver committed:
Bugfix: copy firehose peer set before iterating

carver committed:
TEMP: longer timeout for firehose tests

carver committed:
increase log level of disconnecting peer

carver committed:
Show lag time when importing headers for Beam Sync

carver committed:
Don't crash when peer goes away during stats log

carver committed:
Show subprotocols in debug peer printouts

carver committed:
Beam witness propagation: Skip queuing up empty list of node hashes

Also, show how many node hashes are new, locally. (Which is generally
interesting data!)

carver committed:
Downgrade reliable witness send log to info

carver committed:
Capture and log if peer disappears during Pong

carver committed:
Refuse to beam sync if "caught up" to lagging peer

carver committed:
Beam: don't ask peer for nodes got since queuing

carver committed:
DIRTY SHIT

carver committed:
Quiet a noisy previewing txn log

carver committed:
rebase-fixups

gsalgado commented on trinity/sync/beam/chain.py:
```diff
@@ -142,12 +146,13 @@ def __init__(
         )
 
-        self._backfiller = BeamStateBackfill(db, peer_pool, token=self.cancel_token)
+        self._witness_collector = BeamStateWitnessCollector(db, peer_pool, self.cancel_token)
```
Did you actually intend to replace the backfiller with the witness collector or was this meant to be reverted before merging?

carver replied:
I think this was a kind of a hack. I don't really remember, since it's from ~1 year ago. The basic idea was probably so that we could parallelize the requests for the witness data. But I think that's not so important, based on how many nodes we'll need to collect per block (maybe one or two full `GetNodeData` requests). So it would probably be best to simplify the witness collector a lot, and just trigger "urgent" requests for data.
